state-sync/state-sync-v2/data-streaming-service/src/data_stream.rs

```diff
@@ -308,15 +318,10 @@ impl<T: DiemDataClient + Send + Clone + 'static> DataStream<T> {
 
     /// Notifies the Diem data client of a bad client response
     fn notify_bad_response(&self, data_client_response: &DataClientResponse) {
-        let response_id = data_client_response.response_id;
+        let response_id = data_client_response.id();
         let response_error = ResponseError::InvalidPayloadDataType;
-        let diem_data_client = self.diem_data_client.clone();
-
-        tokio::spawn(async move {
-            let client_response = diem_data_client.notify_bad_response(response_id, response_error);
-            let _ = client_response.await;
-            // TODO(joshlind): log the response if it's an error?
-        });
+        self.diem_data_client
```

Comment:
nit: Why put this on the critical path? I'd prefer to leave it async, e.g., from a client perspective, I don't know how expensive it might be to call this.

Reply:
IMO the data client interface contract should specify that `notify_bad_response` and `get_global_data_summary` are intended to be cheap. i think part of this also involves modifying the `notify_bad_response` API so we don't need to maintain a map from request_id => request context. instead we can probably add that context into the response itself

Reply:
Cool! I was going to suggest documenting this on the API, but you were one step ahead of me 😄